### ISSUE TITLE
Somewhere in the code, causes gradient tracking to break

### ISSUE THREAD

**Issue by jasonrichdarmawan** [seal-rg/recurrent-pretraining#28]
Previous PR: https://github.com/seal-rg/recurrent-pretraining/issues/27
TLDR of Previous PR: The goal is to store hidden state's gradient. The problem is the hidden state's gradient is not stored automatically because it's a non leaf tensor. However, `.retain_grad()` is ignored by PyTorch, resulting in empty hidden state's gradient.

Re-created because the PR cannot be reopened.

~~Initially, I thought it was because of multi-GPUs. But, after creating small and reproducible code, PyTorch is able to track the gradient, even between GPUs.~~

Edit: Tried to re-create `RavenForCausalLM`. It's not looking good when I use `device_map` parameter. The hidden_state gradients will be missing.

Note: Will upload the debugging code soon and reference it here. It's 10pm now, I am going to continue tomorrow.

The code below use `from_config` because it's just a dummy `RavenForCausalLM`, with no `transformer.core_block` and so on. But that's enough to cause the hidden_state gradients to be missing.

```
model = AutoModelForCausalLM.from_config(
  config=config,
)
device_map = {
  "transformer.wte": 0,
  "transformer.prelude": 0,
  "transformer.coda": 1,
  "transformer.ln_f": 1,
  "lm_head": 0,
}
model = dispatch_model(
  model=model,
  device_map=device_map,
)
```

The small and reproducible code which works, to demonstrate that PyTorch can track the gradient between GPUs.
```
# %%

import torch
from torch import Tensor
from jaxtyping import Float

class SandwichBlock(torch.nn.Module):
  def __init__(self, layer_index: int, device: str):
    super().__init__()
    self.layer_index = layer_index
    self.weight = torch.nn.Parameter(
      torch.randn(1, requires_grad=True, device=device)
    )

  def forward(
    self,
    x: Float[Tensor, "size"],
    depth_index: int,
    all_hidden_states: dict[int, Float[Tensor, "size"]]
  ) -> tuple[
    Float[Tensor, "size"], 
    dict[int, Float[Tensor, "size"]]
  ]:
    x = x.to(device=self.weight.device) + self.weight
    
    x.retain_grad()

    all_hidden_states[depth_index] = x

    return x, all_hidden_states

input_embeds = torch.randn(4, requires_grad=True)

all_hidden_states = {}

prelude = torch.nn.ModuleList([
  SandwichBlock(layer_index=0, device="cuda:0"),
  SandwichBlock(layer_index=1, device="cuda:0"),
])
coda = torch.nn.ModuleList([
  SandwichBlock(layer_index=130, device="cuda:1"),
  SandwichBlock(layer_index=131, device="cuda:1"),
])

for block in prelude:
  input_embeds, all_hidden_states = block(
    x=input_embeds, 
    depth_index=block.layer_index, 
    all_hidden_states=all_hidden_states
  )

x = input_embeds.to("cuda:1")

for block in coda:
  x, all_hidden_states = block(
    x=x,
    depth_index=block.layer_index, 
    all_hidden_states=all_hidden_states,
  )

all_hidden_states[0].detach()
"""
If `modified_dict[0].grad` is not `None` even
after `modified_dict[0].detach()` is executed,
it indicates that the `detach()` operation
is not affecting the computation graph in way
you might expect. This behavior can occur because
`.detach()` creates a new tensor that is disconnected
from the computation graph, but it does not
modify the original tensor stored in `modified_dict[0]`.
"""

x.sum().float().backward()

for layer_index, hidden_state in all_hidden_states.items():
  print(f"Layer {layer_index} hidden state grad:", hidden_state.grad)

# %%
```

**Comment by JonasGeiping** [comment]
Regarding Huginn-specific problems, be sure to turn off non-gradient steps: `model.forward`  takes the `num_steps` argument which is parsed as 
``` 
        if num_steps is None:
            num_steps_no_grad, num_steps_with_grad = self.randomized_iteration_sampler()  # type: ignore
        elif hasattr(num_steps, "__len__") and len(num_steps) > 1:
            num_steps_no_grad, num_steps_with_grad = num_steps
        else:
            num_steps_no_grad, num_steps_with_grad = num_steps, torch.tensor(0) if not x.is_meta else 0
``` 
so, if you want to flow gradients through all 32 iterations, you should provide `num_steps=torch.tensor([0, 32])` instead of  `num_steps=torch.tensor([32, 0])` or  `num_steps=32`

**Comment by jasonrichdarmawan** [comment]
> Hi, sorry I'm only reading this now. `device_map` is indeed only designed for inference, and not training, in Hugging Face, see https://huggingface.co/docs/accelerate/concept_guides/big_model_inference#run-the-model. This is a limitation of the Hugging Face implementation of device mapping which detaches the graph when moving across GPUs - this is somewhere in their wrapper.

No need to apologize, in fact, thanks for getting back to me! And apologies on my end, as I initially thought this was a `huginn-0125` bug.

For context: last month I was using [NNsight](https://nnsight.net/tutorials/) with [NDIF](https://ndif.us/) and was able to easily extract hidden state gradients from `deepseek-ai/DeepSeek-R1-Distill-Llama-8B`. I hadn’t realized that `NNsight` uses hooks under the hood, which led me to dig into files like `transformers.models.modeling_llama.py` to figure out why the same approach didn’t work with `huginn-0125`.

On the bright side, I learned a few important things:
1. Why `torch.nn.Module` must return a `tuple` instead of a `dict` (otherwise `torch.nn.Module.register_full_backward_pre_hook` won’t be triggered).
2. Why it's important to have a base model (e.g., `RavenPreTrainedModel`, `RavenModel`) to support task-specific models like `RavenForCausalLM` or `RavenForQuestionAnswering`.

For now, I’m still using `raven_modeling_minimal.py` from `huginn-0125`. Once I get some results, and hopefully more time from my supervisor, I’d love to contribute to this great architecture.

In the meantime, the solution was to use `torch.nn.Module.register_full_backward_pre_hook`, as shown in this [retain_grad example](https://github.com/jasonrichdarmawan/steering-vectors-on-recurrent-model/tree/4ed48832a85d04697f953dfcb14bd6cbeea8e7c9/debugging/retain_grad).

> Regarding Huginn-specific problems, be sure to turn off non-gradient recurrent steps: `model.forward` takes the `num_steps` argument which is parsed as
> 
> ```
>         if num_steps is None:
>             num_steps_no_grad, num_steps_with_grad = self.randomized_iteration_sampler()  # type: ignore
>         elif hasattr(num_steps, "__len__") and len(num_steps) > 1:
>             num_steps_no_grad, num_steps_with_grad = num_steps
>         else:
>             num_steps_no_grad, num_steps_with_grad = num_steps, torch.tensor(0) if not x.is_meta else 0
> ```
> 
> so, if you want to flow gradients through all (e.g. 32) iterations, you should provide `num_steps=torch.tensor([0, 32])` instead of `num_steps=torch.tensor([32, 0])` or `num_steps=32`, as the latter two will parse as `num_steps_no_grad`.

Thank you! I'm using `num_steps=(0, 32)` without wrapping it in `torch.tensor`, and it works on my end, probably because I'm still using `raven_modeling_minimal.py` from `huginn-0125` instead of the newer version.

I am going to close this issue

**Comment by JonasGeiping** [comment]
Sounds good! For posteriority, yeah not wrapping into tensors is fine too, it just needs to an object that can be unpacked.
